### DISCUSSION TITLE
Something about chain and composition

### DISCUSSION THREAD

**Discussion by yurkimus** [ramda/ramda#3445] [Q&A] [answered]
Hello,

Is there a name for such operation presented in a `chain` method? 

```
chain f g x <=> f (g x) x
```

If I need a _reversed chain_, what can I use with _ramda_ and does that composition also have a name?

```
reversed_chain f g x <=> f x (g x)
```

**Comment by adispring** [comment]
`ap`

**Reply by yurkimus** [reply]
Yes, that's what I looked for, thanks.

There is a comment in a REPL: _R.ap can also be used as S combinator_.

Does the chain method has it's related combinator or it's more close to monadic operations?

**Reply by adispring** [reply]
I think `R.chain` can just be consider as a monadic operation(`bind` in Haskell).
`R.ap` can be consider as an [Applicative functor](https://en.wikipedia.org/wiki/Applicative_functor) operation, as Function be consider as an Applicative functor. Also, `R.ap` can be consider as a S combinator.

**Reply by CrossEye** [reply, marked as answer]
Always interesting reading on this subject:

https://gist.github.com/Avaq/1f0636ec5c8d6aed2e45
